leakpro/attacks/mia_attacks/abstract_mia.py

```diff
@@ -256,3 +261,117 @@ def run_attack(self:Self) -> Union[AttackResult, List[AttackResult]]:
 
         """
         pass
+
+    def _memorization(self:Self) -> None:
+        """Run memorization score enhancement.
+
+        Memorization enhances the attack performance by only inlucing vulnerable data points
+
+        """
+
+        self.logger.info("Preparing memorization")
+        if not self.online:
+            self.logger.info("Using the offline version of the attack we make some assumptions in the absence of IN-models")
+
+        if self.signal.__class__.__name__ == "ModelLogits":
+            logits = self.shadow_models_logits
+            target_logits = self.target_logits
+        else:
+            logits_function = ModelLogits()
+            logits = np.swapaxes(logits_function(self.shadow_models, self.audit_data), 0, 1)
+            target_logits = np.swapaxes(logits_function([self.target_model], self.audit_data), 0, 1)
+
+        self.logger.info("Calculating memorization")
+
+        # Initialize memorization score
+        self.memorization_score = np.zeros(len(logits), dtype=float)
+        logits = self.softmax_logits(logits)
+        if self.online:
+            for i, (logit, mask, label) in tqdm(enumerate(zip(logits, self.in_indices_mask, self.audit_data._labels))):
+                self.memorization_score[i] = np.mean(logit[mask, label]) - np.mean(logit[~mask, label])
+        else:
+            for i, (logit, target_logit, label) in tqdm(enumerate(zip(logits, target_logits, self.audit_data._labels))):
+                self.memorization_score[i] = np.mean(target_logit[label]) - np.mean(logit[~mask, label])
```

Comment:
Has this been tested? Does it work?

Reply:
It has been tested, works great in LiRA actually!:)